openraft/src/membership/membership.rs

```diff
@@ -44,12 +56,53 @@ impl Membership {
     pub fn new_single(members: BTreeSet<NodeId>) -> Self {
         let configs = vec![members];
         let all_nodes = Self::build_all_nodes(&configs);
-        Membership { configs, all_nodes }
+        let learners = BTreeSet::new();
+        Membership {
+            learners,
+            configs,
+            all_nodes,
+        }
+    }
+
+    pub fn new_single_with_learners(members: BTreeSet<NodeId>, learners: BTreeSet<NodeId>) -> Self {
+        let configs = vec![members];
+        let all_nodes = Self::build_all_nodes(&configs);
+        Membership {
+            learners,
+            configs,
+            all_nodes,
+        }
     }
 
     pub fn new_multi(configs: Vec<BTreeSet<NodeId>>) -> Self {
         let all_nodes = Self::build_all_nodes(&configs);
-        Membership { configs, all_nodes }
+        let learners = BTreeSet::new();
+        Membership {
+            learners,
+            configs,
+            all_nodes,
+        }
+    }
+
+    pub fn new_multi_with_learners(configs: Vec<BTreeSet<NodeId>>, learners: BTreeSet<NodeId>) -> Self {
+        let all_nodes = Self::build_all_nodes(&configs);
+        Membership {
+            learners,
+            configs,
+            all_nodes,
+        }
+    }
+
+    pub fn add_learner(&mut self, id: &NodeId) {
+        self.learners.insert(*id);
+    }
+
+    pub fn remove_learner(&mut self, id: &NodeId) {
+        self.learners.remove(id);
+    }
+
+    pub fn all_learners(&self) -> &BTreeSet<NodeId> {
+        &self.learners
     }
 
     pub fn all_nodes(&self) -> &BTreeSet<NodeId> {
```

Comment:
`all_nodes()` still returns only members. It should be renamed to `all_members()` or else.